src/dir.targets

```diff
@@ -51,10 +56,25 @@
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets"
           Condition="'$(TargetFrameworkIdentifier)' != '.NETPortable'" />
 
+  <!-- Restore packages -->
+  <PropertyGroup>
+    <PackagesConfig Condition="Exists('$(MSBuildProjectDirectory)\packages.config')">$(MSBuildProjectDirectory)\packages.config</PackagesConfig>
+    <RestorePackages Condition="'$(RestorePackages)' == '' and Exists('$(PackagesConfig)')">true</RestorePackages>
+    <RestorePackagesSemaphore>$(PackagesDir)$(MSBuildProjectFile).package.restored</RestorePackagesSemaphore>
+  </PropertyGroup>
+  <Target Name="RestorePackages" 
+          BeforeTargets="ResolveNuGetPackages"
+          Inputs="$(PackagesConfig)"
+          Outputs="$(RestorePackagesSemaphore)"
+          Condition="'$(RestorePackages)' == 'true'">
+    <Exec Command="$(NugetRestoreCommand) &quot;$(PackagesConfig)&quot;" StandardOutputImportance="Low" />
+    <Touch Files="$(RestorePackagesSemaphore)" AlwaysCreate="true" />
```

Comment:
Wow, so my assertion that restoring packages should be negligible when all packages are already present is clearly invalid. Given that, I can accept this optimization; but I wonder if we can separately look into why restoring packages takes so long when there shouldn't be anything to do?

I am looking into RunTestsForProject and CopyTestToTestDirectory.